ORDER BY NEWID() を実行できるようにする

sd-tomita opened the pull request:
## 概要
`Sdx.Db.Sql.Select` では 現在は `ORDER BY {column} {ASC | DESC}` が可能ですが、
現状は必ずカラムの指定が必須 + ASC/DESC が付いてきてしまいます。
`ORDER BY NEWID()` のような SQL文は指定できないので、このPRで対応します。
とりあえずオーダーランダムのみ対応します。

### メモ

MySQLでいうところの `ORDER BY RAND()` を SQL server で実現する方法
http://stackoverflow.com/questions/19412/how-to-request-a-random-row-in-sql#answer-19419

sd-tomita committed:
wip

sd-tomita commented:
### テスト実行方法

https://github.com/SunriseDigital/cs-sdx/pull/16#issuecomment-136242220

sd-tomita commented:
## 想定使用コード

以下のように使えると、以前のフレームワーク使用時とほぼ同じ感覚で使用できるので良さそうな気がします。

``` C#
var main = Sdx.Db.Adapter.Manager.Get("main");
var db = main.Read;
var select = db.CreateSelect();

select
  .AddFrom("Shop")
  .AddOrder("contract")
  .AddOrderRand()
  .Where.Add("shop_id", "someShopId");

using (var conn = db.CreateConnection())
{
  conn.Open();
  this.RecordSet = conn.FetchRecordSet(select);
}
```

gomo commented:
``` c#
    public Context AddOrderRandom()
    {
      var column = new Column(Expr.Wrap(Adapter.RandomOrderKeyword), Name);
      Select.OrderList.Add(column);

      return this;
    }
```

sd-tomita commented:
うーん、おしい。

``` SQL
SELECT * FROM [Shop] WHERE [Shop].[shop_id] = @0 ORDER BY NEWID() ASC
@0 = 1394
```

sd-tomita committed:
Order に Null を許容する記述を追加

ORDER BY NEWID() を行う場合、ASC・DESCどちらも付けないようにする
必要があったため。

sd-tomita committed:
AddOrderRandom を試験実装

とりあえず動作チェックのため、NEWID()を直書きで実装してみました。

sd-tomita committed:
Adapter.RandomOrderKeyword を追加

sd-tomita committed:
Select への追加漏れ

gomo commented:
``` c#
if(column.Order != null)
{
  var str = ((Order)column.Order).SqlString();
}
```

sd-tomita committed:
BuildSelectString 内でキャストする方法に変更

sd-tomita committed:
ユニットテストをコミット

これだけではちょっと心許ない気がするが、ひとまず
無いよりはまだいいのでコミットしています。

Adapter.RandomOrderKeyword を internal にしていなければ
ORDER BY + select.Adapter.RandomOrderKeyword という方法もありかも。

gomo commented on UnitTest/Sdx/Db/Sql/Select.cs:
```diff
@@ -2040,5 +2040,25 @@ private void RunCountRow(TestDb testDb)
+      select
+        .AddFrom(new Test.Orm.Table.Shop())
+        .AddOrderRandom();
+      Assert.True(select.Build().CommandText.Contains("ORDER BY"));
```
このテストは微妙かも。

https://github.com/SunriseDigital/cs-sdx/blob/da93a395c6bb2ed3e518d5d7b758a24102e0ed30/Sdx/Db/Adapter/Base.cs#L170

ここをPublicにしないとダメだからこうしたんだと思うけど、getのみだったら問題無いのでちゃんとテストしたほうがいいと思った。ちゃんと正しいSQLが生成されてるか確認しましょう。Containsっていうも微妙だと思った。

SQLサーバーとMySQLでエスケープ記号が違うのでテスト難しいけど
https://github.com/SunriseDigital/cs-sdx/blob/da93a395c6bb2ed3e518d5d7b758a24102e0ed30/UnitTest/Sdx/Db/Sql/Select.cs#L1786-L1790
こんな感じでできます。また、このメソッドは改行やスペース（一番最初以外）を空に置換するので見やすく書けます。

😇

sd-tomita replied:
> https://github.com/SunriseDigital/cs-sdx/blob/da93a395c6bb2ed3e518d5d7b758a24102e0ed30/Sdx/Db/Adapter/Base.cs#L170
> 
> ここをPublicにしないとダメだからこうしたんだと思うけど

おっしゃる通りです。
public にして、Test 側でも使うようにしてみます。

> SQLサーバーとMySQLでエスケープ記号が違うのでテスト難しいけど
> https://github.com/SunriseDigital/cs-sdx/blob/da93a395c6bb2ed3e518d5d7b758a24102e0ed30/UnitTest/Sdx/Db/Sql/Select.cs#L1786-L1790
> こんな感じでできます。また、このメソッドは改行やスペース（一番最初以外）を空に置換するので見やすく書けます。

ありがとうございます！
参考にSQL全文が期待した内容に一致しているかをテストするように変更してみます。

sd-tomita committed:
ユニットテスト内でも使用したかったので public に変更

sd-tomita committed:
ユニットテストを修正

ORDER BY が含まれているかではテストとしては不十分だったので
期待しているSQLが完全に生成されているかSQL文を全て見るように変更。

gomo commented:
👍 
Good!

ピボタルに貼ったリンクですけど、気を利かせてくれたんだと思いますけど、PRのリンクがあれば十分ですよ。`New changes since you last viewed`があるので。

逆に貼ってもらったリンクからPRに行こうと思ったら、いけなくて開き直しました 😎

sd-tomita commented:
> ピボタルに貼ったリンクですけど、気を利かせてくれたんだと思いますけど、PRのリンクがあれば十分ですよ。`New changes since you last viewed`があるので。
> 
> 逆に貼ってもらったリンクからPRに行こうと思ったら、いけなくて開き直しました 😎 

おお、失礼しました。たしかにURL見たらPR配下じゃなくなっていました。
次回以降はふつうにPRのリンク貼るようにします。

sd-tomita committed:
Merge branch 'master' into proj/Select/AddOrderRandom
